synthesizer/process/src/lib.rs

```diff
@@ -123,28 +154,103 @@ impl<N: Network> Process<N> {
+            .unique_by(|(id, _)|*id) // don't add duplicates.
+            .filter(|(program_id, _)| program_id != &ProgramID::<N>::from_str("credits.aleo").unwrap()) // don't add the credits.aleo stack.
+            .filter(|(program_id, _)| !stacks.contains(program_id)) // don't add stacks present in the cache.
+            .collect::<Vec<_>>();
+        // Determine the required capacity.
+        let current_capacity = stacks.cap().get().saturating_sub(stacks.len());
+        let mut required_capacity = programs_to_add.len().saturating_sub(current_capacity);
+        if required_capacity > 0 {
+            debug!("Evicting {required_capacity} stacks from the cache.");
+        }
+        // If the new stacks require more capacity, attempt to remove the least recently used stacks.
+        while required_capacity > 0 {
+            // To avoid dangling stacks, we only remove stacks that do not have more than 1 reference (e.g. as external stack).
+            // For this not to loop endlessly, we assume to always quickly find stacks with just one reference, so:
+            // 1. queried stacks must be short-lived, which is essential anyway to avoid memory leaks.
+            // 2. queried stacks must expire before a querying thread queries the next stack.
+            let root_program_ids = stacks
+                .iter()
+                .rev()
+                .filter_map(|(program_id, stack)| (std::sync::Arc::<_>::strong_count(stack) == 1).then_some(program_id))
+                .take(required_capacity)
+                .cloned()
+                .collect::<Vec<_>>();
+            // Lower the required capacity.
+            required_capacity = required_capacity.saturating_sub(root_program_ids.len());
+            // Evict the old stacks from the cache.
+            for program_id in root_program_ids {
+                if stacks.pop(&program_id).is_none() {
+                    bail!("Could not find expected program id in cache.")
+                }
+            }
+        }
+        // Add a new stacks into the cache.
+        for (program_id, stack) in programs_to_add {
+            stacks.put(program_id, stack);
+        }
+        Ok(())
+    }
+
+    /// Returns the size of the cache.
+    #[inline]
+    pub fn num_stacks(&self) -> usize {
+        self.stacks.lock().len()
     }
 }
 
 impl<N: Network> Process<N> {
+    /// Initializes a new process.
+    /// Assumption: this is only called in test code.
+    #[inline]
+    pub fn load_testing_only() -> Result<Self> {
+        Process::load_from_storage(Some(aleo_std::StorageMode::Development(0)))
+    }
+
     /// Initializes a new process.
     #[inline]
-    pub fn load() -> Result<Self> {
-        let timer = timer!("Process::load");
+    pub fn load_from_storage(storage_mode: Option<StorageMode>) -> Result<Self> {
+        let timer = timer!("Process::load_from_storage");
+
+        debug!("Opening storage");
```

Comment:
this is a bit too vague to me if I were to see it in logs at random; perhaps we can specify the context a bit more or remove this and the following `debug` log if they're no longer needed for dev purposes

Reply:
Removing them